src/main/java/org/apache/mesos/offer/AttributeStringUtils.java

```diff
@@ -0,0 +1,110 @@
+package org.apache.mesos.offer;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.StringJoiner;
+import java.util.StringTokenizer;
+
+import org.apache.mesos.Protos.Attribute;
+import org.apache.mesos.Protos.Value;
+
+/**
+ * Tools for converting {@link Attribute}s into strings. Spec is defined by Mesos documentation at:
+ * http://mesos.apache.org/documentation/latest/attributes-resources/
+ */
+public class AttributeStringUtils {
+
+    private static final String ATTRIBUTE_LIST_SEPARATOR = ";";
+    private static final char ATTRIBUTE_KEYVAL_SEPARATOR = ':';
+
+    private AttributeStringUtils() {
+        // do not instantiate
+    }
+
+    public static List<String> toStringList(String joinedAttributes) {
+        StringTokenizer tokenizer = new StringTokenizer(joinedAttributes, ATTRIBUTE_LIST_SEPARATOR);
+        List<String> tokens = new ArrayList<>();
+        while (tokenizer.hasMoreTokens()) {
+            tokens.add(tokenizer.nextToken());
+        }
+        return tokens;
+    }
+
+    /**
+     * Converts the provided list of zero or more attributes into a string suitable for parsing by
+     * {@link #parseString(String)}.
+     *
+     * @throws IllegalArgumentException if some part of the provided attributes couldn't be
+     * serialized
+     */
+    public static String toString(List<Attribute> attributes) throws IllegalArgumentException {
+        StringJoiner joiner = new StringJoiner(ATTRIBUTE_LIST_SEPARATOR);
+        for (Attribute attribute : attributes) {
+            joiner.add(toString(attribute));
+        }
+        return joiner.toString();
+    }
+
+    /**
+     * Converts the provided attribute into a string which follows the format defined by Mesos:
+     * <code>
+     * attributes : attribute ( ";" attribute )*
+     * attribute : text ":" ( scalar | range | text )
+     * text : [a-zA-Z0-9_/.-]
+     * scalar : floatValue
+     * floatValue : ( intValue ( "." intValue )? ) | ...
+     * intValue : [0-9]+
+     * range : "[" rangeValue ( "," rangeValue )* "]"
+     * rangeValue : scalar "-" scalar
+     * set : "{" text ( "," text )* "}"
+     * </code>
+     *
+     * NOTE that it is difficult if not impossible to consistently perform the inverse of this
+     * operation. For example, how can you tell if something is supposed to be a SCALAR value or a
+     * TEXT value? [0-9.]+ is valid in both cases! Your best hope is to consistently convert to
+     * string, and then convert strings...
+     *
+     * @throws IllegalArgumentException if some part of the provided attributes couldn't be
+     * serialized
+     */
+    public static String toString(Attribute attribute) throws IllegalArgumentException {
+        StringBuffer buf = new StringBuffer();
+        buf.append(attribute.getName());
+        buf.append(ATTRIBUTE_KEYVAL_SEPARATOR);
+        switch (attribute.getType()) {
+        case RANGES: {
```

Comment:
ehh.. they're pretty small as it is, and share the same key/KEYVAL_SEPARATOR handling